File: javascript/ping.js
/* PING — Minimal reference implementation in JavaScript.
 *
 * Fixes vs draft 0.1:
 *  - Outbox now stays open until ACKNOWLEDGED (not CONTENT_PULLED).
 *  - Sender advances to PING_RECEIVED when Blog 2 confirms receipt.
 *  - advance() helper is used throughout instead of manual string assignment.
 *  - Retry loop with exponential backoff on the initial ping.
 *  - Blocking: receiver maintains a blocklist; blocked pings are silently
 *    dropped (sender cannot distinguish blocked from offline).
 *  - Tiered endpoints: /ping/public (open) and /ping/user (token-gated).
 *  - GET /inbox lets the receiver list pending pings.
 *
 * Out of scope for this draft (spec §Non-Goals / future drafts):
 *  - Anonymous proxy tier (requires external infrastructure).
 *  - Private tier / E2EE.
 *  - Message masks.
 *  - P2P / DHT content transfer.
 */

const express = require("express");
const crypto = require("crypto");
const app = express();
app.use(express.json());

// CORS — required for the browser-based demo (ping-demo.html)
app.use((req, res, next) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Methods', 'GET,POST,DELETE');
  res.header('Access-Control-Allow-Headers', 'Content-Type,Authorization');
  if (req.method === 'OPTIONS') return res.sendStatus(200);
  next();
});

// ---------------------------------------------------------------------------
// State machine
// ---------------------------------------------------------------------------

const State = Object.freeze({
  PING_EXISTS:    "PING_EXISTS",
  PING_RECEIVED:  "PING_RECEIVED",
  CONTENT_PULLED: "CONTENT_PULLED",
  ACKNOWLEDGED:   "ACKNOWLEDGED",
});

const TRANSITIONS = {
  [State.PING_EXISTS]:    State.PING_RECEIVED,
  [State.PING_RECEIVED]:  State.CONTENT_PULLED,
  [State.CONTENT_PULLED]: State.ACKNOWLEDGED,
};

function advance(state) {
  const next = TRANSITIONS[state];
  if (!next) throw new Error(`${state} is a terminal state`);
  return next;
}

// ---------------------------------------------------------------------------
// Storage (in-memory — replace with durable store for production)
// ---------------------------------------------------------------------------

const outbox   = new Map(); // sender side:   pingId → entry
const inbox    = new Map(); // receiver side: pingId → entry
const blocklist = new Set(); // blocked origins (domain or IP string)

// ---------------------------------------------------------------------------
// Retry / backoff
// ---------------------------------------------------------------------------

// Intervals in ms: 5 s, 30 s, 2 min, 10 min, 1 h
const RETRY_INTERVALS_MS = [5_000, 30_000, 120_000, 600_000, 3_600_000];

// Fires in the background; advances outbox entry to PING_RECEIVED on success.
async function attemptPingWithRetry(pingId, target, payload) {
  for (let attempt = 0; attempt <= RETRY_INTERVALS_MS.length; attempt++) {
    const entry = outbox.get(pingId);
    // Bail if entry was removed or already advanced past PING_EXISTS externally.
    if (!entry || entry.state !== State.PING_EXISTS) return;

    try {
      const resp = await fetch(`${target}/ping/public`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(10_000),
      });
      if (resp.ok) {
        entry.state = advance(entry.state); // PING_EXISTS → PING_RECEIVED
        return;
      }
    } catch (_) {
      // Blog 2 is offline or unreachable; try again after backoff.
    }

    const wait = RETRY_INTERVALS_MS[attempt];
    if (wait === undefined) break; // exhausted retry schedule
    await new Promise((r) => setTimeout(r, wait));
  }
  // State remains PING_EXISTS after all retries; caller may inspect and decide.
}

// ---------------------------------------------------------------------------
// Sender endpoints
// ---------------------------------------------------------------------------

// POST /send — Blog 1 creates a new ping and begins the retry loop.
app.post("/send", (req, res) => {
  const { origin, target, payload_type = "ask", content = "", content_hash } = req.body;
  if (!origin || !target) {
    return res.status(400).json({ error: "origin and target are required" });
  }

  const pingId = crypto.randomUUID();
  outbox.set(pingId, {
    state: State.PING_EXISTS,
    origin,
    target,
    payload_type,
    content,
    content_hash,
  });

  // Kick off the retry loop without blocking the response.
  const payload = { ping_id: pingId, origin, payload_type, content_hash };
  attemptPingWithRetry(pingId, target, payload).catch(() => {});

  res.status(201).json({ ping_id: pingId, state: State.PING_EXISTS });
});

// GET /outbox/:pingId — Blog 2 pulls content from here.
// Content MUST remain available until ACKNOWLEDGED (spec §Sender Records).
app.get("/outbox/:pingId", (req, res) => {
  const entry = outbox.get(req.params.pingId);
  if (!entry) return res.status(404).end();

  // Only close the outbox once the receiver has acknowledged receipt.
  if (entry.state === State.ACKNOWLEDGED) return res.status(410).end();

  // Advance to CONTENT_PULLED on first pull (idempotent on subsequent pulls
  // before ACK, so a failed ACK allows the receiver to retry).
  if (entry.state !== State.CONTENT_PULLED) {
    entry.state = State.CONTENT_PULLED;
  }

  res.json({
    ping_id: req.params.pingId,
    payload_type: entry.payload_type,
    content: entry.content,
  });
});

// POST /outbox/:pingId/ack — Blog 2 signals it has successfully pulled content.
// Idempotent: re-acknowledging an already-acknowledged entry is a no-op.
app.post("/outbox/:pingId/ack", (req, res) => {
  const entry = outbox.get(req.params.pingId);
  if (!entry) return res.status(404).end();

  if (entry.state !== State.ACKNOWLEDGED) {
    entry.state = State.ACKNOWLEDGED;
  }

  res.status(204).end();
});

// ---------------------------------------------------------------------------
// Receiver — helpers
// ---------------------------------------------------------------------------

// Registers a received ping in the inbox, or silently drops if origin is blocked.
// Silence is the canonical response to an unwanted ping (spec §Blocking and Consent):
// the sender cannot distinguish blocked from offline.
function receivePing(ping_id, origin, payload_type, content_hash, res) {
  if (blocklist.has(origin)) {
    // Return 204 so the sender cannot detect the block via an error code.
    return res.status(204).end();
  }

  inbox.set(ping_id, {
    state: State.PING_RECEIVED,
    origin,
    payload_type,
    content_hash,
  });

  res.status(204).end();
}

// ---------------------------------------------------------------------------
// Receiver — ping endpoints (access tiers)
// ---------------------------------------------------------------------------

// Public tier — open to any sender.
app.post("/ping/public", (req, res) => {
  const { ping_id, origin, payload_type, content_hash } = req.body;
  receivePing(ping_id, origin, payload_type, content_hash, res);
});

// User tier — requires a valid bearer token.
// Replace VALID_TOKENS with real credential storage.
const VALID_TOKENS = new Set(["change-me-secret-token"]);

app.post("/ping/user", (req, res) => {
  const auth = req.headers["authorization"] ?? "";
  const token = auth.replace(/^Bearer\s+/i, "").trim();
  if (!VALID_TOKENS.has(token)) return res.status(401).end();

  const { ping_id, origin, payload_type, content_hash } = req.body;
  receivePing(ping_id, origin, payload_type, content_hash, res);
});

// ---------------------------------------------------------------------------
// Receiver — pull and inbox management
// ---------------------------------------------------------------------------

// POST /pull/:pingId — Blog 2 initiates a content pull from Blog 1's outbox.
app.post("/pull/:pingId", async (req, res) => {
  const entry = inbox.get(req.params.pingId);
  if (!entry) return res.status(404).end();
  if (entry.state === State.ACKNOWLEDGED) return res.status(410).end();

  // Fetch content from sender.
  let resp;
  try {
    resp = await fetch(`${entry.origin}/outbox/${req.params.pingId}`, {
      signal: AbortSignal.timeout(15_000),
    });
  } catch (err) {
    return res.status(502).json({ error: "could not reach sender" });
  }

  if (!resp.ok) {
    return res.status(502).json({ error: "pull failed", upstream: resp.status });
  }

  const content = await resp.json();
  entry.state = advance(entry.state); // PING_RECEIVED → CONTENT_PULLED
  entry.content = content.content;

  // Acknowledge to sender. If the ACK fails, state stays at CONTENT_PULLED
  // so the receiver can retry the pull (which re-sends the ACK) without loss.
  try {
    const ack = await fetch(`${entry.origin}/outbox/${req.params.pingId}/ack`, {
      method: "POST",
      signal: AbortSignal.timeout(10_000),
    });
    if (ack.ok) {
      entry.state = advance(entry.state); // CONTENT_PULLED → ACKNOWLEDGED
    }
  } catch (_) {
    // ACK failed; pull succeeded. Caller may re-POST /pull to retry the ACK.
  }

  res.json(content);
});

// GET /inbox — list pings that have not yet been acknowledged.
app.get("/inbox", (_req, res) => {
  const pending = [];
  for (const [pingId, entry] of inbox) {
    if (entry.state !== State.ACKNOWLEDGED) {
      pending.push({
        ping_id: pingId,
        state: entry.state,
        origin: entry.origin,
        payload_type: entry.payload_type,
        content_hash: entry.content_hash,
      });
    }
  }
  res.json(pending);
});

// ---------------------------------------------------------------------------
// Receiver — blocking
// ---------------------------------------------------------------------------

// POST /block — add an origin to the blocklist.
app.post("/block", (req, res) => {
  const { origin } = req.body;
  if (!origin) return res.status(400).json({ error: "origin is required" });
  blocklist.add(origin);
  res.status(204).end();
});

// DELETE /block — remove an origin from the blocklist.
app.delete("/block", (req, res) => {
  const { origin } = req.body;
  if (!origin) return res.status(400).json({ error: "origin is required" });
  blocklist.delete(origin);
  res.status(204).end();
});

// GET /block — inspect the current blocklist (admin/debug use).
app.get("/block", (_req, res) => {
  res.json({ blocked: [...blocklist] });
});

// ---------------------------------------------------------------------------
// Start
// ---------------------------------------------------------------------------

const port = process.argv[2] ?? 5000;
app.listen(port, () => console.log(`PING listening on :${port}`));

File: python/ping.py
"""PING — Reference implementation in Python.

Aligned with Protocol Specification Draft 0.1.

State machine:
    PING_EXISTS → PING_RECEIVED → CONTENT_PULLED → ACKNOWLEDGED

Key changes from initial demo:
  - advance() is now the sole mechanism for state transitions everywhere.
  - Sender-side outbox state advances to PING_RECEIVED once Blog 2 confirms receipt.
  - Background retry/backoff loop delivers pings to an offline Blog 2.
  - Hash verification: Blog 2 rejects content whose SHA-256 doesn't match the ping.
  - /inbox lets Blog 2 inspect pending pings before deciding whether to pull.
  - /ignore/<ping_id> implements the silent-ignore consent path.
  - /outbox/<ping_id> returns 410 only after ACKNOWLEDGED (not after CONTENT_PULLED).
  - /outbox/<ping_id>/ack is idempotent — safe to call on an already-terminal entry.
  - Locks guard all shared state so the retry thread and Flask threads don't race.
"""

import hashlib
import sys
import threading
import time
from enum import Enum
from uuid import uuid4

from flask import Flask, request, jsonify
import requests as http

app = Flask(__name__)


# ---------------------------------------------------------------------------
# State machine
# ---------------------------------------------------------------------------

class State(Enum):
    PING_EXISTS    = "PING_EXISTS"
    PING_RECEIVED  = "PING_RECEIVED"
    CONTENT_PULLED = "CONTENT_PULLED"
    ACKNOWLEDGED   = "ACKNOWLEDGED"

TRANSITIONS = {
    State.PING_EXISTS:    State.PING_RECEIVED,
    State.PING_RECEIVED:  State.CONTENT_PULLED,
    State.CONTENT_PULLED: State.ACKNOWLEDGED,
}

def advance(state: State) -> State:
    """Return the next state, or raise if already terminal."""
    if state not in TRANSITIONS:
        raise ValueError(f"{state.value} is a terminal state")
    return TRANSITIONS[state]


# ---------------------------------------------------------------------------
# In-memory storage
# ---------------------------------------------------------------------------

# Sender side:  ping_id → {state, origin, target, payload_type, content,
#                           content_hash}
outbox: dict = {}
outbox_lock = threading.Lock()

# Receiver side: ping_id → {state, origin, payload_type, content_hash,
#                            content (set after pull)}
inbox: dict = {}
inbox_lock = threading.Lock()

# Ping IDs that this instance has explicitly ignored — silently dropped on
# future re-delivery attempts.
ignored: set = set()


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------

def sha256(text: str) -> str:
    return hashlib.sha256(text.encode()).hexdigest()


def _log(msg: str) -> None:
    print(msg, flush=True)


# ---------------------------------------------------------------------------
# Retry / backoff delivery loop (sender side)
# ---------------------------------------------------------------------------

# Delays in seconds between successive retry attempts.
RETRY_SCHEDULE = [2, 5, 10, 30, 60, 300]


def _retry_worker(ping_id: str) -> None:
    """Background thread: keep attempting ping delivery until Blog 2 confirms."""
    for delay in RETRY_SCHEDULE:
        time.sleep(delay)

        with outbox_lock:
            entry = outbox.get(ping_id)
            if entry is None or entry["state"] != State.PING_EXISTS:
                _log(f"[retry] {ping_id}: already past PING_EXISTS, stopping")
                return

        _log(f"[retry] {ping_id}: attempting delivery to {entry['target']}")
        try:
            resp = http.post(
                f"{entry['target']}/ping",
                json={
                    "ping_id":      ping_id,
                    "origin":       entry["origin"],
                    "payload_type": entry["payload_type"],
                    "content_hash": entry["content_hash"],
                },
                timeout=5,
            )
            if resp.status_code == 204:
                with outbox_lock:
                    entry["state"] = advance(entry["state"])  # → PING_RECEIVED
                _log(f"[retry] {ping_id}: delivered → PING_RECEIVED")
                return
            else:
                _log(f"[retry] {ping_id}: remote returned {resp.status_code}, will retry")
        except http.RequestException as exc:
            _log(f"[retry] {ping_id}: network error ({exc}), will retry")

    _log(f"[retry] {ping_id}: exhausted retry schedule, state remains PING_EXISTS")


# ---------------------------------------------------------------------------
# Sender endpoints
# ---------------------------------------------------------------------------

@app.post("/send")
def send_ping():
    """Blog 1: compose content, record it in the outbox, and notify Blog 2."""
    body = request.json
    ping_id  = str(uuid4())
    origin   = body["origin"]
    target   = body["target"]
    content  = body.get("content", "")

    # If no hash is supplied, compute one so Blog 2 can verify integrity.
    content_hash = body.get("content_hash") or sha256(content)

    with outbox_lock:
        outbox[ping_id] = {
            "state":        State.PING_EXISTS,
            "origin":       origin,
            "target":       target,
            "payload_type": body.get("payload_type", "ask"),
            "content":      content,
            "content_hash": content_hash,
        }

    # Attempt immediate delivery.
    delivered = False
    try:
        resp = http.post(
            f"{target}/ping",
            json={
                "ping_id":      ping_id,
                "origin":       origin,
                "payload_type": outbox[ping_id]["payload_type"],
                "content_hash": content_hash,
            },
            timeout=5,
        )
        if resp.status_code == 204:
            with outbox_lock:
                # Advance sender-side state now that Blog 2 has confirmed receipt.
                outbox[ping_id]["state"] = advance(outbox[ping_id]["state"])  # → PING_RECEIVED
            delivered = True
    except http.RequestException:
        pass

    if not delivered:
        # Blog 2 is offline — start the background retry loop.
        t = threading.Thread(target=_retry_worker, args=(ping_id,), daemon=True)
        t.start()

    state_value = outbox[ping_id]["state"].value
    _log(f"[send] {ping_id}: initial state → {state_value}")
    return jsonify({"ping_id": ping_id, "state": state_value}), 201


@app.get("/outbox/<ping_id>")
def serve_content(ping_id):
    """Blog 2 pulls content from here.

    Returns 410 only after ACKNOWLEDGED (terminal). Re-pulls during
    CONTENT_PULLED are allowed — they handle the case where Blog 2's ack
    request was lost in transit.
    """
    with outbox_lock:
        entry = outbox.get(ping_id)
        if not entry:
            return "", 404

        if entry["state"] == State.ACKNOWLEDGED:
            return "", 410  # terminal — outbox entry is closed

        # Allow serving from PING_RECEIVED or CONTENT_PULLED (idempotent pull).
        entry["state"] = State.CONTENT_PULLED
        return jsonify({
            "ping_id":      ping_id,
            "payload_type": entry["payload_type"],
            "content":      entry["content"],
            "content_hash": entry["content_hash"],
        })


@app.post("/outbox/<ping_id>/ack")
def receive_ack(ping_id):
    """Blog 2 confirms delivery. Idempotent — safe to call on an already-terminal entry."""
    with outbox_lock:
        entry = outbox.get(ping_id)
        if not entry:
            return "", 404
        if entry["state"] != State.ACKNOWLEDGED:
            entry["state"] = advance(entry["state"])  # CONTENT_PULLED → ACKNOWLEDGED
            _log(f"[ack] {ping_id}: → ACKNOWLEDGED")
    return "", 204


# ---------------------------------------------------------------------------
# Receiver endpoints
# ---------------------------------------------------------------------------

@app.post("/ping")
def receive_ping():
    """Accept a ping notification from a remote blog.

    Silently drops pings from origins on the ignored list.
    """
    body    = request.json
    ping_id = body["ping_id"]

    with inbox_lock:
        if ping_id in ignored:
            _log(f"[ping] {ping_id}: in ignored set, dropping silently")
        else:
            inbox[ping_id] = {
                "state":        State.PING_RECEIVED,
                "origin":       body["origin"],
                "payload_type": body["payload_type"],
                "content_hash": body.get("content_hash"),
            }
            _log(f"[ping] {ping_id}: recorded → PING_RECEIVED")

    # Always return 204 — silence reveals nothing to the sender.
    return "", 204


@app.get("/inbox")
def list_inbox():
    """Return all non-terminal inbox entries so Blog 2 can decide what to pull.

    This is the consent inspection point: Blog 2 sees origin and payload type
    before committing to a pull.
    """
    with inbox_lock:
        items = [
            {
                "ping_id":      pid,
                "origin":       e["origin"],
                "payload_type": e["payload_type"],
                "content_hash": e["content_hash"],
                "state":        e["state"].value,
            }
            for pid, e in inbox.items()
            if e["state"] != State.ACKNOWLEDGED
        ]
    return jsonify(items)


@app.post("/pull/<ping_id>")
def pull_content(ping_id):
    """Blog 2 decides to fetch content — this is where consent is exercised.

    Verifies the SHA-256 hash of the received content against the hash
    advertised in the original ping. Returns 422 on mismatch.
    """
    with inbox_lock:
        entry = inbox.get(ping_id)
    if not entry:
        return jsonify({"error": "unknown ping_id"}), 404
    if entry["state"] == State.ACKNOWLEDGED:
        return jsonify({"error": "already acknowledged"}), 409

    # Fetch content from the sender's outbox.
    try:
        resp = http.get(f"{entry['origin']}/outbox/{ping_id}", timeout=10)
    except http.RequestException as exc:
        return jsonify({"error": f"pull failed: {exc}"}), 502

    if resp.status_code != 200:
        return jsonify({"error": f"sender returned {resp.status_code}"}), 502

    payload = resp.json()

    # Hash verification — reject tampered or corrupted content.
    advertised_hash = entry["content_hash"]
    if advertised_hash:
        actual_hash = sha256(payload["content"])
        if actual_hash != advertised_hash:
            _log(f"[pull] {ping_id}: hash mismatch — expected {advertised_hash}, got {actual_hash}")
            return jsonify({
                "error":    "hash mismatch",
                "expected": advertised_hash,
                "actual":   actual_hash,
            }), 422

    with inbox_lock:
        entry["state"]   = advance(entry["state"])  # PING_RECEIVED → CONTENT_PULLED
        entry["content"] = payload["content"]

    # Acknowledge delivery to the sender.
    try:
        ack = http.post(f"{entry['origin']}/outbox/{ping_id}/ack", timeout=5)
        if ack.status_code == 204:
            with inbox_lock:
                entry["state"] = advance(entry["state"])  # CONTENT_PULLED → ACKNOWLEDGED
            _log(f"[pull] {ping_id}: → ACKNOWLEDGED")
    except http.RequestException as exc:
        _log(f"[pull] {ping_id}: ack failed ({exc}), state left at CONTENT_PULLED")

    return jsonify(payload)


@app.post("/ignore/<ping_id>")
def ignore_ping(ping_id):
    """Blog 2 silently ignores a ping.

    Removes it from the inbox and marks the ID so future re-delivery attempts
    are also dropped. Blog 1 receives no notification — silence is the
    canonical response to an unwanted ping.
    """
    with inbox_lock:
        inbox.pop(ping_id, None)
        ignored.add(ping_id)
    _log(f"[ignore] {ping_id}: silently ignored")
    return "", 204


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 5000
    # threaded=True so Flask can handle concurrent requests while the retry
    # loop is running in the background.
    app.run(port=port, threaded=True)
